tests/withBlockTracker.ts

```diff
@@ -0,0 +1,252 @@
+import util from 'util';
+import SafeEventEmitter from '@metamask/safe-event-emitter';
+import { JsonRpcRequest, JsonRpcResponse } from 'json-rpc-engine';
+import { Provider } from '../src/BaseBlockTracker';
+import {
+  SubscribeBlockTracker,
+  SubscribeBlockTrackerOptions,
+} from '../src/SubscribeBlockTracker';
+import {
+  PollingBlockTracker,
+  PollingBlockTrackerOptions,
+} from '../src/PollingBlockTracker';
+
+interface WithPollingBlockTrackerTypes {
+  withBlockTrackerCallback: (args: {
+    provider: FakeProvider;
+    blockTracker: PollingBlockTracker;
+  }) => void | Promise<void>;
+  withBlockTrackerOptions: {
+    provider?: FakeProviderOptions;
+    blockTracker?: PollingBlockTrackerOptions;
+  };
+}
+
+interface WithSubscribeBlockTrackerTypes {
+  withBlockTrackerCallback: (args: {
+    provider: FakeProvider;
+    blockTracker: SubscribeBlockTracker;
+  }) => void | Promise<void>;
+  withBlockTrackerOptions: {
+    provider?: FakeProviderOptions;
+    blockTracker?: SubscribeBlockTrackerOptions;
+  };
+}
+
+type FakeProviderStub =
+  | {
+      methodName: string;
+      response: { result: any } | { error: string };
+    }
+  | {
+      methodName: string;
+      implementation: () => void;
+    }
+  | {
+      methodName: string;
+      error: string;
+    };
+
+interface FakeProviderOptions {
+  stubs?: FakeProviderStub[];
+}
+
+/**
+ * FakeProvider is an implementation of the provider that a subclass of
+ * BaseBlockTracker takes, supporting the same expected methods with the same
+ * expected interface, except that fake responses for the various RPC methods
+ * that the provider supports can be supplied. This ends up easier to define
+ * than using `jest.spyOn(...).mockImplementationOnce(...)` due to the types
+ * that `sendAsync` takes.
+ */
+class FakeProvider extends SafeEventEmitter implements Provider {
+  private stubs: FakeProviderStub[];
+
+  private originalStubs: FakeProviderStub[];
+
+  constructor({ stubs = [] }: FakeProviderOptions = {}) {
```

Comment:
Nit: It'd be great to have a JSDoc comment here explaining what these options do.

Reply:
Sure. Added in [c3e20c9](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/c3e20c989f39fad72cc1bc2f002839d6db4f61cd).